### PR TITLE
Add Oracle Bulk Datatype tests

### PR DESCRIPTION
**Pull request by dhwanilpatel**
Add oracle bulk datatype ITs.

### CODE REVIEW

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java [hunk 1]
```diff
@@ -0,0 +1,2912 @@
...
+   *       simulating a mapping of textual CHAR bounds locally to standard Spanner BYTES (like
+   *       RPAD('A', 1000)) strictly pass `'A'` strings cross-boundary. When the standard bytes
+   *       engine receives Avro strings meant for BYTES, it defaults to attempting native Hex
+   *       decoding (`Hex.decodeHex(...)`). Instead of failing format validation, literal Character
+   *       `'A'` coincidentally represents valid Hex (`0xAA`). A string of 1000 inserted `'A'`
+   *       string characters gracefully executes dynamically into exactly 500 contiguous bytes of
+   *       hexadecimal `0xAA`! Native Spanner Base64 rendering dynamically maps `0xAA` identically
+   *       out precisely into the `"qqqq..qo="` strings!
+   *   <li><b>Unsafe Plaintext -> BYTES Mappings (`DecoderException` Fallback Removal)</b>: Because
+   *       Datastream natively passes binary Oracle types (`RAW`, `BLOB`) directly mapped as
```
Comment (bharadwaj-aditya) [inline v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java:90]: why is datastream relevant in this test ?

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java [hunk 2]
```diff
@@ -0,0 +1,2912 @@
...
+        continue;
+      }
+      if (tableName.endsWith("_to_bytea_table") && !tableName.equals("raw_to_bytea_table")) {
+        continue;
+      }
+      if (entry.getValue().isEmpty()) {
+        System.out.println("Skipping entirely emptied bound test for: " + tableName);
+        continue;
+      }
+      System.out.println("VERIFYING TABLE: " + tableName);
```
Comment (bharadwaj-aditya) [inline v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java:148]: we should not have sysouts in the tests. 

Can you also update the skill to ensure there are no sysouts in the tests.

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java [hunk 3]
```diff
@@ -0,0 +1,2912 @@
...
+                put("varchar2_col", "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA...");
+                put("id", 4L);
+              }
+            }));
+    expectedData.put(
+        "varchar2_to_string_table",
+        java.util.Arrays.asList(
+            new java.util.HashMap<String, Object>() {
+              {
+                put("varchar2_col", " ");
```
Comment (bharadwaj-aditya) [inline v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java:191]: This seems like an very verbose way of writing these tests. 

Is this how the tests are written in the other sources ?
using something like Map.of or direct put methods will reduce these blocks to 1 line each.

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java [hunk 4]
```diff
@@ -0,0 +1,2912 @@
...
...gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICA=");
```
Comment (bharadwaj-aditya) [inline v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java:417]: seems like approx the same row is written thrice. What is the difference ?

**Review by bharadwaj-aditya — COMMENTED** [review]
The tests seem to have been written in an extremely verbose manner. Can you look into make these more manageable ?

File: v2/sourcedb-to-spanner/src/test/resources/oracle/OracleDataTypesIT/oracle-schema.sql [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+INSERT INTO "raw_to_bytes_table" ("id", "raw_col") VALUES (3, UTL_RAW.CAST_TO_RAW('A'));
+INSERT INTO "raw_to_bytes_table" ("id", "raw_col") VALUES (4, UTL_RAW.CAST_TO_RAW('DROP TABLE'));
+CREATE TABLE "raw_to_varchar_base64_table" ("id" NUMBER PRIMARY KEY, "raw_col" RAW(2000));
+INSERT INTO "raw_to_varchar_base64_table" ("id", "raw_col") VALUES (1, NULL);
+INSERT INTO "raw_to_varchar_base64_table" ("id", "raw_col") VALUES (2, NULL);
+INSERT INTO "raw_to_varchar_base64_table" ("id", "raw_col") VALUES (3, UTL_RAW.CAST_TO_RAW('A'));
+INSERT INTO "raw_to_varchar_base64_table" ("id", "raw_col") VALUES (4, UTL_RAW.CAST_TO_RAW('DROP TABLE'));
+CREATE TABLE "long_raw_table" ("id" NUMBER PRIMARY KEY, "long_raw_col" LONG RAW);
+INSERT INTO "long_raw_table" ("id", "long_raw_col") VALUES (1, UTL_RAW.CAST_TO_RAW('""'));
+INSERT INTO "long_raw_table" ("id", "long_raw_col") VALUES (2, UTL_RAW.CAST_TO_RAW('"A"*100000'));
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/resources/oracle/OracleDataTypesIT/oracle-schema.sql:547]: We can test exact string here and will test the longer strings or limits in load test.

File: v2/sourcedb-to-spanner/src/test/resources/oracle/OracleDataTypesPGDialectIT/oracle-schema.sql [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+INSERT INTO "raw_to_bytea_table" ("id", "raw_col") VALUES (3, UTL_RAW.CAST_TO_RAW('A'));
+INSERT INTO "raw_to_bytea_table" ("id", "raw_col") VALUES (4, UTL_RAW.CAST_TO_RAW('DROP TABLE'));
+CREATE TABLE "raw_to_varchar_base64_table" ("id" NUMBER PRIMARY KEY, "raw_col" RAW(2000));
+INSERT INTO "raw_to_varchar_base64_table" ("id", "raw_col") VALUES (1, NULL);
+INSERT INTO "raw_to_varchar_base64_table" ("id", "raw_col") VALUES (2, NULL);
+INSERT INTO "raw_to_varchar_base64_table" ("id", "raw_col") VALUES (3, UTL_RAW.CAST_TO_RAW('A'));
+INSERT INTO "raw_to_varchar_base64_table" ("id", "raw_col") VALUES (4, UTL_RAW.CAST_TO_RAW('DROP TABLE'));
+CREATE TABLE "long_raw_table" ("id" NUMBER PRIMARY KEY, "long_raw_col" LONG RAW);
+INSERT INTO "long_raw_table" ("id", "long_raw_col") VALUES (1, UTL_RAW.CAST_TO_RAW('""'));
+INSERT INTO "long_raw_table" ("id", "long_raw_col") VALUES (2, UTL_RAW.CAST_TO_RAW('"A"*100000'));
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/resources/oracle/OracleDataTypesPGDialectIT/oracle-schema.sql:546]: We can test exact string here and will test the longer strings or limits in load test.

File: v2/sourcedb-to-spanner/src/test/resources/oracle/OracleDataTypesIT/oracle-schema.sql [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+INSERT INTO "smallint_to_float64_table" ("id", "smallint_col") VALUES (2, -922337203685477);
+INSERT INTO "smallint_to_float64_table" ("id", "smallint_col") VALUES (3, 0);
+INSERT INTO "smallint_to_float64_table" ("id", "smallint_col") VALUES (4, NULL);
+INSERT INTO "smallint_to_float64_table" ("id", "smallint_col") VALUES (5, 922337203685476);
+CREATE TABLE "smallint_pk_table" ("smallint_pk_col" SMALLINT PRIMARY KEY);
+INSERT INTO "smallint_pk_table" ("smallint_pk_col") VALUES (922337203685476);
+INSERT INTO "smallint_pk_table" ("smallint_pk_col") VALUES (-922337203685477);
+INSERT INTO "smallint_pk_table" ("smallint_pk_col") VALUES (0);
+CREATE TABLE "date_table" ("id" NUMBER PRIMARY KEY, "date_col" DATE);
+INSERT INTO "date_table" ("id", "date_col") VALUES (1, TIMESTAMP '0001-01-01 00:00:00');
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/resources/oracle/OracleDataTypesIT/oracle-schema.sql:491]: I haven't seen the flakiness for this as we uses the self hosted github runner where timezone would be fixed and wont change.

File: v2/sourcedb-to-spanner/src/test/resources/oracle/OracleDataTypesPGDialectIT/oracle-schema.sql [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+INSERT INTO "smallint_to_double_precision_table" ("id", "smallint_col") VALUES (2, -922337203685477);
+INSERT INTO "smallint_to_double_precision_table" ("id", "smallint_col") VALUES (3, 0);
+INSERT INTO "smallint_to_double_precision_table" ("id", "smallint_col") VALUES (4, NULL);
+INSERT INTO "smallint_to_double_precision_table" ("id", "smallint_col") VALUES (5, 922337203685476);
+CREATE TABLE "smallint_pk_table" ("smallint_pk_col" SMALLINT PRIMARY KEY);
+INSERT INTO "smallint_pk_table" ("smallint_pk_col") VALUES (922337203685476);
+INSERT INTO "smallint_pk_table" ("smallint_pk_col") VALUES (-922337203685477);
+INSERT INTO "smallint_pk_table" ("smallint_pk_col") VALUES (0);
+CREATE TABLE "date_table" ("id" NUMBER PRIMARY KEY, "date_col" DATE);
+INSERT INTO "date_table" ("id", "date_col") VALUES (1, TIMESTAMP '0001-01-01 00:00:00');
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/resources/oracle/OracleDataTypesPGDialectIT/oracle-schema.sql:490]: I haven't seen the flakiness for this as we uses the self hosted github runner where timezone would be fixed and wont change.

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java (hunk 2 above)
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java]: Removed sys outs. Will update skill in seperate PR.

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java (hunk 3 above)
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java]: Tried reducing this by using Map.of(). It has significantly reduced the size of the code.

> Is this how the tests are written in the other sources ?

Other tests also has similar way of creating expacted data, but that seems to be already optimised with single line inserts.

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java (hunk 4 above)
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java]: Those are for testing different bytes limits in spanner. Also similar data has been tested for various other alternatives leading to duplicate. 

Instead of hardcoding long string, I have updated it to generate the hex string for verification. It has reduced the size of code a lot.

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java (hunk 1 above)
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java]: Actually limitation is in AvroToValueMapper. I was discussing on how this works on live template with this, and seems like agent mentioned those information as well in here.

Updated Java doc.

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java [hunk 9]
```diff
@@ -0,0 +1,978 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.templates.oracle;
+
+import static org.apache.beam.it.truthmatchers.PipelineAsserts.assertThatResult;
```
Comment (bharadwaj-aditya) [inline v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java:18]: future usage suggestion - we should be able to use the all data types data set across all our templates.

**Review by bharadwaj-aditya — APPROVED** [review]
LGTM

File: v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java (hunk 9 above)
Comment (dhwanilpatel) [reply v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/oracle/OracleDataTypesIT.java:18]: Yes, We can evaluate doing that for one source across all templates, for across source schema might be different.
